crates/sui-core/src/authority_server.rs

```diff
@@ -540,6 +542,198 @@ impl ValidatorService {
+            protocol_config.enable_soft_bundle(),
+            SuiError::FailedToSubmitToConsensus("SoftBundle: Feature not enabled".into()).into()
+        );
+        let max_num_certs = protocol_config.soft_bundle_max_num_certificates() as usize;
+        let max_serialized_size = protocol_config.soft_bundle_max_serialized_size_bytes() as usize;
+        fp_ensure!(
+            certificates.len() > 0,
+            SuiError::FailedToSubmitToConsensus("SoftBundle: Bundle size must not be zero".into())
+                .into()
+        );
+        fp_ensure!(
+            certificates.len() <= max_num_certs,
+            SuiError::FailedToSubmitToConsensus(
+                "SoftBundle: Bundle contains too many certificates".into()
+            )
+            .into()
+        );
+
+        // Basic checks
+        let mut total_serialized_size: usize = 0;
+        let mut last_gas_price = None;
+        for certificate in &certificates {
+            // CRITICAL! Validators should never sign an external system transaction.
+            fp_ensure!(
+                !certificate.is_system_tx(),
+                SuiError::InvalidSystemTransaction.into()
+            );
+            fp_ensure!(
+                certificate.contains_shared_object(),
+                SuiError::FailedToSubmitToConsensus(
+                    "SoftBundle: All certificate must contain at least one shared objects".into()
+                )
+                .into()
+            );
+            let gas_price = certificate.data().intent_message().value.gas_price();
+            if let Some(last_gas_price) = last_gas_price {
+                fp_ensure!(
+                    gas_price == last_gas_price,
+                    SuiError::FailedToSubmitToConsensus(
+                        "SoftBundle: All certificate must have the same gas price".into()
+                    )
+                    .into()
+                );
+            }
+            last_gas_price = Some(gas_price);
+
+            certificate
+                .data()
+                .validity_check(epoch_store.protocol_config())?;
+            total_serialized_size += certificate.data().serialized_size()?;
+        }
+        fp_ensure!(
+            total_serialized_size <= max_serialized_size,
+            SuiError::FailedToSubmitToConsensus("SoftBundle: Bundle size too large".into()).into()
+        );
+
+        let mut responses = Vec::new();
+        responses.resize(certificates.len(), None);
+        let _metrics_guard = metrics.submit_certificate_consensus_latency.start_timer();
+        let mut cont = Vec::new();
+
+        // First pass
+        let mut cert_verify_futures = Vec::new();
+        for (request_idx, certificate) in certificates.into_iter().enumerate() {
+            // 1) Check if cert already executed
+            let tx_digest = *certificate.digest();
+            let already_executed =
+                match state.get_signed_effects_and_maybe_resign(&tx_digest, &epoch_store)? {
```

Comment:
can you please use `state.is_tx_already_executed()`, as this function may potentially create a new unused BLS signature (expensive)